chainer/device_resident.py

```diff
@@ -260,17 +289,17 @@ def _device_to_method_name_and_kwargs(self, device):
 
     def visit_array(self, arr):
         assert isinstance(arr, chainer.get_array_types())
-        if not (self._skip_between_cupy_devices
-                and self._device.xp is cuda.cupy
-                and isinstance(arr, cuda.ndarray)):
+        skip = _warn_to_gpu(
+            self._device, arr, legacy=self._skip_between_cupy_devices)
+        if not skip:
```

Comment:
It's a bit hard to understand this code.
How about this:
```py
        if (self._skip_between_cupy_devices
                and self._device.xp is cuda.cupy
                and isinstance(arr, cuda.ndarray)):
            return _warn_to_gpu(arr, self._device)  # note: argument reverted because it's more natural
        return self._device.send(arr)

Reply:
How about moving the condition outside?
I feel the current implementation is still difficult to understand.

Reply:
Specifically, in the current implementation the condition is hidden (from the calling code) and not clear what (and when) `_warn_to_gpu` does.